### DISCUSSION TITLE
Getting handle on router added globally in a store with a "setup function"

### DISCUSSION THREAD

**Discussion by demyte** [vuejs/pinia#1717] [Help and Questions] [answered]
Hi all,

I have setup my router in the store globally as showed in the docs:

```typescript
import type { Router } from 'vue-router'

export interface PiniaCustomProperties {
    router: Router
  }
}

const pinia = createPinia()

pinia.use(({ store }) => {
  store.router = markRaw(router)
})
```

And my store is being setup with a setup function:
```typescript
export const useGlobalStore = defineStore('global', () => {

// This leaves router as undefined
const router = useRouter()  

})
```

My google-fu is failing me trying to work out how to get a handle on that router instance that was injected as I need to use it because `const router = useRouter()` is undefined.

Thanks :)

**Comment by posva** [comment, marked as answer]
One could attach the router instance to the pinia instance and retrieve through the _active pinia_:

```ts
import { getActivePinia } from 'pinia'

const pinia = createPinia()
pinia.router = router // extend the Pinia Interface for typings

const useRouter = () => getActivePinia().router as Router
```

You can use this `useRouter()` version within setup stores.

In Nuxt, since you don't manually create the router or the pinia, you can add a module or plugin to set the router to the pinia instance

---

Note that if you are not doing SSR, you can just directly import the `router` variable from your router file.

**Reply by demyte** [reply]
Thanks for the response.

I tried the code you suggested, but `getActivePinia().router` is `undefined` on use, but this put me on the right track to change it to be:

``` typescript
export const useDirectRouter = () => markRaw(router)
```

And then used it right next to where I needed it:

``` typescript
const router = useDirectRouter()
router.push({
  name: 'Home',
  force: true,
})
```

If I used it higher in the store setup function, I get the following message because of the sequence.

![image](https://user-images.githubusercontent.com/48166/195585530-93112394-bcbd-4db8-b765-1c49eba2086a.png)

Thanks again for taking the time to respond :)

**Reply by posva** [reply]
The useRouter is showed you **must be used within setup stores**. Your version is the same as importing the router directly, if you are not doing SSR, you should do that, it's the simplest way to get the router instance 🙂

**Reply by demyte** [reply]
I started by using the version you put in as per below but it is coming up as undefined - am I missing something?  (I am not doing SSR)

Definition when creating the pinia store
![image](https://user-images.githubusercontent.com/48166/195593347-38198aa0-fb76-41cb-b0aa-0d38836ea6ea.png)

Imported inside the store setup
![image](https://user-images.githubusercontent.com/48166/195593263-b414b26d-fa99-4296-9760-0901d9c6002f.png)

In use:
![image](https://user-images.githubusercontent.com/48166/195593744-179676d4-17b1-4e7b-90f7-978fa17a582c.png)

**Reply by posva** [reply]
yes, you are not adding the `router` to the pinia instance, you are adding it to stores 😅

**Reply by demyte** [reply]
ahh.. I see my error with missing `pinia.router = router`... thanks 👍

When I do it that way, I still get an error:

![image](https://user-images.githubusercontent.com/48166/195716240-965d97f6-26eb-4dcf-a17c-9aa3aa759fff.png)

So, I put it in an anonymous function it works:

![image](https://user-images.githubusercontent.com/48166/195716712-a5f9d473-ea99-45a9-9fb8-22377b435d2b.png)

But I still have to use it right where it is used, not in the top of the store setup - I can live with that.

Thanks again for helping this Pinia novice.

**Reply by posva** [reply]
Follow the error: you have to do it after both, pinia and the router are initialized, do it in the `main.ts` file
